geomstats/learning/em_expectation_maximization.py

```diff
@@ -0,0 +1,265 @@
+DEFAULT_TOL = 1e-2
+ZETA_LOWER_BOUND = 5e-2
+ZETA_UPPER_BOUND = 2.
+ZETA_STEP = 0.001
+PDF_TOL = 1e-15
+SUM_CHECK_PDF = 1e-4
+MEAN_MAX_ITER = 150
+
+
+class RiemannianEM():
+    """Class for running Expectation-Maximisation."""
+
+    def __init__(self,
+                 riemannian_metric,
+                 n_gaussian=8,
+                 initialisation_method='random',
+                 tol=DEFAULT_TOL,
+                 mean_method='default',
+                 point_type='vector'):
+        """Expectation-maximization algorithm on Poincaré Ball.
+
+        A class for performing Expectation-Maximization on the
+        Poincaré Ball to fit data into a Gaussian Mixture Model.
+
+        Parameters
+        ----------
+        riemannian_metric : object of class RiemannianMetric
+        The geomstats Riemmanian metric associated with
+                            the used manifold
+
+        n_gaussian : int
+        Number of Gaussian components in the mix
+
+        initialisation_method : basestring
+        Choice between initialization method for variances, means and weights
+               'random' : will select random uniformally train point as
+                         initial centroids
+
+                #TODO: implement kmeans initialisation
+                'kmeans' : will apply Riemannian kmeans to deduce
+                variances and means that the EM will use initially
+
+        tol : float
+            Convergence factor. If the difference of mean distance
+            between two step is lower than tol
+
+        mean_method: basestring
+            Specify the method to compute the mean.
+
+        point_type: basestring
+            Specify whether to use vector or matrix representation
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        self.n_gaussian = n_gaussian
+        self.riemannian_metric = riemannian_metric
+        self.initialisation_method = initialisation_method
+        self.tol = tol
+        self.mean_method = mean_method
+        self.point_type = point_type
+        self._dimension = None
+        self.mixture_coefficients = None
+        self.variances = None
+        self.means = None
+        self.normalization_factor = None
+
+    def update_posterior_probabilities(self, posterior_probabilities):
+        """Posterior probabilities update function."""
+        self.mixture_coefficients = gs.mean(posterior_probabilities, 0)
+
+    def update_means(self, data, posterior_probabilities,
+                     lr_means, tau_means, max_iter=DEFAULT_MAX_ITER):
+        """Means update function."""
+        n_gaussian = posterior_probabilities.shape[-1]
+
+        mean = FrechetMean(
+            metric=self.riemannian_metric,
+            method=self.mean_method,
+            lr=lr_means,
+            tau=tau_means,
+            max_iter=max_iter,
+            point_type=self.point_type)
+
+        data_expand = gs.expand_dims(data, 1)
```

Comment:
We might not need all this reshaping with the next vectorization PRs, but unfortunately they are not merged yet! To come soon!

Reply:
I saw some changes with vectorization affecting the distance/log/exp, but I still need the reshapes/repeats/expands for the moment because I have situations foo(a,b) with a.shape = (m,n) and b.shape = (x,n)  with m different than x, and expecting to apply for each couple (m,x) the foo function.

Reply:
Got it! This answer might apply to my comment above too.